### PR TITLE
Guide for monitoring tests run by CI - final push

### PR DESCRIPTION
**Pull request by RobertKielty**
Re-submission of the **Test Monitoring Guide** written by @alejandrox1

We will apply feedback as requested by @justaugustus [here](https://github.com/kubernetes/community/pull/5244#issuecomment-770019562) on the original PR 

[Origin thread on Slack](https://github.com/kubernetes/community/pull/5244#issuecomment-770019562rl) 

@spiffxp @thejoycekung @hasheddan @jeremyrickard @justaugustus

### CODE REVIEW

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:152]: I made a slightly annotated version of this image so that we don't have to keep repeating "the value above that" or "the value below that" and it's a little easier for the "Since when has it been failing" section. 

image

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:163]:
```suggestion
At 15:07 EDT, the test
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:170]:
```suggestion
If one or both of these tests continue failing, or if they fail frequently
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:225]:
```suggestion
using https://go.k8s.io/triage.
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:239]: Ahhh, I see the comment I made about TestGrid tab names not being "real" job names has come up here. Not sure whether this should be moved further up to step 1?? but it requires the reader to know to click on the cells in TestGrid for Spyglass which we walk through in step 4?

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:244]:
```suggestion
Then we can use this full job name in Triage: https://go.k8s.io/triage?pr=1&job=ci-kubernetes-node-kubelet-features.

At the time of this writing we saw the following
```

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [reply contributors/devel/sig-testing/monitoring.md:114]: My preference would be to explicitly mention kubernetes/kubernetes

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [reply contributors/devel/sig-testing/monitoring.md:79]: It might cover it? 

I'm not a fan of the wording overall but I won't block on it.  For instance, I'm not sure "I happened to be browsing through" is relevant to a heading of "I see a TestGrid alert".  Maybe a separate section that describes browsing and when/how to act?

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [reply contributors/devel/sig-testing/monitoring.md:53]: We've got some verbiage to that effect here if you want to link for more info: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/flaky-tests.md#flaky-tests

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [inline contributors/devel/sig-testing/monitoring.md:130]: There is overlap here with https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/flaky-tests.md#github-issues-for-known-flakes, ultimately we should converge on one location for instructions on how to file an issue, and link to it.  I don't have strong opinions on whether that's this doc, the other doc, the issue template, or some other location.   But I have strong opinions that we avoid N different guides.

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [reply contributors/devel/sig-testing/monitoring.md:134]: Strongly agree.  The job is _not_ the name of the tab.  It is really important to distinguish the two, because you need the job name to:
- find job config in kubernetes/test-infra
- look at recent runs in https://prow.k8s.io
- look up failures in https://go.k8s.io/triage
- look up top flakes in http://storage.googleapis.com/k8s-metrics/flakes-latest.json

I really think a glossary or explainer up top is needed:
- job name: e.g. `ci-kubernetes-foo`, shows up in prow.k8s.io, is the name of the gcs bucket where results are stored, etc
- testgrid tab name: 'build', is a shortened and possibly non-unique name that is well understand given the context of the dashboard in which it lives

How to translate from one to the other? Quickest way as a human is to click on a cell to see results in prow.k8s.io's spyglass, it will list the job name up top

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [inline contributors/devel/sig-testing/monitoring.md:161]: Note: this is not a guaranteed column heading for all release-blocking jobs.  In fact, it might not be relevant at all when these jobs are migrated away from bootstrap (https://github.com/kubernetes/test-infra/issues/20650)

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [reply contributors/devel/sig-testing/monitoring.md:170]: What does "frequently enough" mean?

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [inline contributors/devel/sig-testing/monitoring.md:182]:
```suggestion
* (Optionally) test-infra commit
```

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [reply contributors/devel/sig-testing/monitoring.md:239]: Echoing what I said above, this should practically be at the top, under a linkable heading like "Glossary" or "Terminology" or something

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [inline contributors/devel/sig-testing/monitoring.md:75]: I'm not sure if it makes sense in this doc, but here's something I do, and see others do, which might be just tribal vs. documented knowledge.

If a test fails on my PR, and it's clearly not related to the code I wrote, I try to find any open issues that appear related (have the name of the test in them, describe a similar error, etc.).  I'll link that in the comment I use to retrigger jobs, either calling the job out specifically:
```
./test pull-kubernetes-foo
https://github.com/kubernetes/kubernetes/issues/foo
```
or even if just spamming retest
```
./retest
https://github.com/kubernetes/kubernetes/issues/foo
```

I might even back-link from the issue to the PR that encountered it, to bump the issue's last updated date.

That way, the issue will accumulate links to other occurrences of the flake.  Lots of links = good signal this is actual pain contributors are regularly experiencing.

File: contributors/devel/sig-testing/monitoring.md
Comment (spiffxp) [inline contributors/devel/sig-testing/monitoring.md:5]: nit: would it make sense to drop this so we can un-indent everything else?

**Review by spiffxp — CHANGES_REQUESTED** [review]
Agree with @thejoycekung's review, made a few suggestions

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:60]: That is an assumption that is being made in this doc and you you are correct to say that it has not been stated. Though I think it might be better to state this up top.

Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:60]: I added an Overview up top to add the "changes made to Kubernetes" context

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:75]: Good stuff, I'll add this.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:134]: Should we report this as a User Ex1perience bug in TestGrid? 

What is the justification for having having the tab name be different from the job name? Seems needlessly confusing to me. 

Also Jorge called this UX pothole out towards the end of the doc.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:152]: Very nice, thank you :) 
Will use this.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:170]: Removing.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:239]: Or reported as a UX defect.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:47]: Nice catch, I made this consistent by taking on your feedback.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:264]: This is brilliant TYSM for bringing this to my attention. I will commit this work separately.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:182]: Removed "Frequently enough"
Rewrote that list to use @thejoycekung 's red-number-annotated TestGrid diagram (Thanks again JK!)

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md:328]: I changed the links to make use of the reference style. 

Thanks again for this suggestion. Much appreciated.

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:216]:
```suggestion
[k8s.io] NodeProblemDetector [NodeFeature:NodeProblemDetector] [k8s.io] SystemLogMonitor should generate node condition and events for corresponding errors [ubuntu]
```
Putting it on 2 lines _really_ makes it look like 2 different tests

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:226]:
```suggestion
[k8s.io] NodeProblemDetector [NodeFeature:NodeProblemDetector] [k8s.io] SystemLogMonitor should generate node condition and events for corresponding errors [cos-stable2]
```
Same comment as before.

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:243]:
```suggestion
- 1: The starting times for each Prow job; each column on the grid
  represents a single run of the Prow job.
- 2: The Prow job run id number.
- 3: The `kubernetes/kubernetes` commit id that was tested.
- 4: These are most commonly the `kubernetes/test-infra` commit ids that were used to build and
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:257]:
```suggestion
community; as the issue reporter you do not have to find the reason for failure
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:261]:
```suggestion
Click on the failed runs (the red cells in the grid) to see the results in 
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:266]: nit: is there a reason why this image (and the next) are included with Markdown notation whereas the ones before used HTML? this should be consistent

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:283]:
```suggestion
You can reference a specific line in the logs by clicking on the line number and
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:297]:
```suggestion
One important detail is that the job names you see on TestGrid are
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:302]:
```suggestion
the job name `ci-kubernetes-node-kubelet-features` can be found at the top left corner of the Spyglass page (notice the "ci-kubernetes-" prefix).
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:305]:
```suggestion
the job field]. Note that the Triage query can be bookmarked and can be used as a
```

File: contributors/devel/sig-testing/monitoring.md
Comment (thejoycekung) [inline contributors/devel/sig-testing/monitoring.md:312]:
```suggestion
**Note**: You can also improve your query by filtering or excluding
```

**Review by thejoycekung — COMMENTED** [review]
Been meaning to look through the latest set of changes for a bit. Thanks for all your work so far @RobertKielty !

**Review by spiffxp — APPROVED** [review]
/approve
/hold
I think we're about at merge and iterate so I'm going to approve. 

But I defer to @thejoycekung for the final say, who continues to raise valid suggestions. Remove hold when you're ready to lgtm and thanks for the thorough reviews!

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md]: TestGrid has been improved on this point. (I think thanks to work done by @spiffxp ! Thank you Aaron) So I'll update the screen shot and adjust the copy accordingly.

Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md]: I've added an svg version of the screen shot. SVG files render fine on github markdown 

I created the svg file using Inkscape (quirky UX but powerful and worth learning effort) 

Inkscape diagrams have layers. Screenshot from TG is in it's own layer. The annotations I added pointing out TG UI elements (commit sources, Prow Job numbers, Prow Job config and desc) are in a separate layer. 

This means that we can reuse those annotations in other screen shots and also will make the screen shot a little more maintainable when we pick up TestGrid improvements in the future.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md]: I don't get what you're saying here, can you please elaborate? 
Do you mean dropping the TOC altogether?

Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md]: I've updated the TOC. Pls review.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md]: Done.

File: contributors/devel/sig-testing/monitoring.md
Comment (RobertKielty) [reply contributors/devel/sig-testing/monitoring.md]: Done

**Comment by ramrodo** [comment]
> But I defer to @thejoycekung for the final say, who continues to raise valid suggestions. Remove hold when you're ready to lgtm and thanks for the thorough reviews!

Hi @thejoycekung, doing a following of this PR, do you have more suggestions or this PR is ready to be merged?

**Comment by thejoycekung** [comment]
hey @ramrodo thanks for following up! I remember chatting with @RobertKielty about this a few weeks ago -- from what I can tell there are still a few unaddressed comments, but once those are fixed I'd love to push this through :)

**Comment by justaugustus** [comment]
Looks like this got stuck.
Let's please take any outstanding review comments as follow-ups.

/lgtm
/hold cancel

**Pull request merged**
